sql/expression/function/convert_tz.go

```diff
@@ -0,0 +1,185 @@
+
+import (
+	"errors"
+	"fmt"
+	"regexp"
+	"time"
+
+	"github.com/dolthub/go-mysql-server/sql"
+)
+
+var offsetRegex = regexp.MustCompile(`(?m)^(\+|\-)(\d{2}):(\d{2})$`) // (?m)^\+|\-(\d{2}):(\d{2})$
+
+type ConvertTz struct {
+	dt     sql.Expression
+	fromTz sql.Expression
+	toTz   sql.Expression
+}
+
+var _ sql.FunctionExpression = (*ConvertTz)(nil)
+
+// NewConvertTz returns an implementation of the CONVERT_TZ() function.
+func NewConvertTz(ctx *sql.Context, dt, fromTz, toTz sql.Expression) sql.Expression {
+	return &ConvertTz{
+		dt:     dt,
+		fromTz: fromTz,
+		toTz:   toTz,
+	}
+}
+
+// Resolved implements the sql.Expression interface.
+func (c *ConvertTz) Resolved() bool {
+	return c.dt.Resolved() && c.fromTz.Resolved() && c.toTz.Resolved()
+}
+
+// String implements the sql.Expression interface.
+func (c *ConvertTz) String() string {
+	return fmt.Sprintf("CONVERT_TZ(%s, %s, %s)", c.dt, c.fromTz, c.toTz)
+}
+
+// Type implements the sql.Expression interface.
+func (c *ConvertTz) Type() sql.Type {
+	return sql.Datetime
+}
+
+// IsNullable implements the sql.Expression interface.
+func (c *ConvertTz) IsNullable() bool {
+	return true
+}
+
+// Eval implements the sql.Expression interface.
+func (c *ConvertTz) Eval(ctx *sql.Context, row sql.Row) (interface{}, error) {
+	from, err := c.fromTz.Eval(ctx, row)
+	if err != nil {
+		return nil, err
+	}
+
+	to, err := c.toTz.Eval(ctx, row)
+	if err != nil {
+		return nil, err
+	}
+
+	dt, err := c.dt.Eval(ctx, row)
+	if err != nil {
+		return nil, err
+	}
+
+	// If either the date, or the timezones/offsets are not correct types we return NULL.
+	datetime, err := sql.Datetime.ConvertWithoutRangeCheck(dt)
+	if err != nil {
+		return nil, nil
+	}
+
+	// Get the time in UTC.
+	datetime = time.Date(datetime.Year(), datetime.Month(), datetime.Day(), datetime.Hour(), datetime.Minute(), datetime.Second(), datetime.Nanosecond(), time.UTC)
+
+	fromStr, ok := from.(string)
+	if !ok {
+		return nil, nil
+	}
+
+	toStr, ok := to.(string)
+	if !ok {
+		return nil, nil
+	}
+
+	converted := convertTimeZone(datetime, fromStr, toStr)
+	if !converted.IsZero() {
+		return sql.Datetime.ConvertWithoutRangeCheck(converted)
+	}
+
+	// If we weren't successful converting by timezone try converting via offsets.
+	converted = convertOffsets(datetime, fromStr, toStr)
+	if converted.IsZero() {
+		return nil, nil
+	}
+
+	return sql.Datetime.ConvertWithoutRangeCheck(converted)
+}
+
+// convertTimeZone returns the conversion of t from timezone fromLocation to toLocation.
+func convertTimeZone(datetime time.Time, fromLocation string, toLocation string) time.Time {
+	fLoc, err := time.LoadLocation(fromLocation)
+	if err != nil {
+		return time.Time{}
+	}
+
+	tLoc, err := time.LoadLocation(toLocation)
+	if err != nil {
+		return time.Time{}
+	}
+
+	getCopy := func(t time.Time, loc *time.Location) time.Time {
+		return time.Date(t.Year(), t.Month(), t.Day(), t.Hour(), t.Minute(), t.Second(), t.Nanosecond(), loc).UTC()
+	}
```

Comment:
I'd make this a function in the module namespace rather than one created inside of this function. I'm assuming this was done to avoid cluttering the namespace, so you could always make it a function of `*ConvertTz`, e.g.:
```go
func (c *ConvertTz) getCopy(t time.Time, loc *time.Location) time.Time {
	return time.Date(t.Year(), t.Month(), t.Day(), t.Hour(), t.Minute(), t.Second(), t.Nanosecond(), loc).UTC()
}
```

Reply:
Not sure if it needs to be a method on ConvertTz. I'll take this out of the other function though